lib/reline/face.rb

```diff
@@ -0,0 +1,157 @@
+      yellow: 43,
+      blue: 44,
+      magenta: 45,
+      cyan: 46,
+      white: 47,
+      bright_black: 100,
+      gray: 100,
+      bright_red: 101,
+      bright_green: 102,
+      bright_yellow: 103,
+      bright_blue: 104,
+      bright_magenta: 105,
+      bright_cyan: 106,
+      bright_white: 107,
+    },
+    style: {
+      reset: 0,
+      bold: 1,
+      faint: 2,
+      italicized: 3,
+      underlined: 4,
+      slowly_blinking: 5,
+      blinking: 5,
+      rapidly_blinking: 6,
+      negative: 7,
+      concealed: 8,
+      crossed_out: 9
+    }
+  }.freeze
+
+  class Config
+    ESSENTIAL_DEFINE_NAMES = %i(default enhanced scrollbar).freeze
+    RESET_SGR = "\e[0m".freeze
+
+    def initialize(name, &block)
+      @definition = {}
+      block.call(self)
+      ESSENTIAL_DEFINE_NAMES.each do |name|
+        @definition[name] ||= { style: :reset, escape_sequence: RESET_SGR }
+      end
+    end
+
+    attr_reader :definition
+
+    def define(name, **values)
+      values[:escape_sequence] = format_to_sgr(values.to_a).freeze
+      @definition[name] = values
+    end
+
+    def [](name)
+      @definition.dig(name, :escape_sequence) or raise ArgumentError, "unknown face: #{name}"
+    end
+
+    private
+
+    def sgr_rgb(key, value)
+      return nil unless rgb_expression?(value)
+      case key
+      when :foreground
+       "38;2;"
+      when :background
+        "48;2;"
+      end + value[1, 6].scan(/../).map(&:hex).join(";")
+    end
+
+    def format_to_sgr(ordered_values)
+      sgr = "\e[" + ordered_values.map do |key_value|
+        key, value = key_value
+        case key
+        when :foreground, :background
+          case value
+          when Symbol
+            SGR_PARAMETERS[key][value]
+          when String
+            sgr_rgb(key, value)
+          end
+        when :style
+          [ value ].flatten.map do |style_name|
+            SGR_PARAMETERS[:style][style_name]
+          end.then do |sgr_parameters|
+            sgr_parameters.include?(nil) ? nil : sgr_parameters
+          end
+        end.then do |rendition_expression|
+          unless rendition_expression
+            raise ArgumentError, "invalid SGR parameter: #{value.inspect}"
+          end
+          rendition_expression
+        end
+      end.join(';') + "m"
+      sgr == RESET_SGR ? RESET_SGR : RESET_SGR + sgr
+    end
+
+    def rgb_expression?(color)
+      color.respond_to?(:match?) and color.match?(/\A#[0-9a-fA-F]{6}\z/)
+    end
+  end
+
+  private_constant :SGR_PARAMETERS, :Config
+
+  def self.[](name)
+    @configs[name]
+  end
+
+  def self.config(name, &block)
+    @configs ||= {}
+    @configs[name] = Config.new(name, &block)
+  end
+
+  def self.configs
+    @configs.transform_values(&:definition)
+  end
+
+  def self.load_initial_configs
+    config(:default) do |conf|
+      conf.define :default, style: :reset
+      conf.define :enhanced, style: :reset
+      conf.define :scrollbar, style: :reset
+    end
```

Comment:
What is this config required for? It does not appear to be used.
Sorry if I have missed it.

Reply:
In IRB, `configs[:default]` is used for all character output except the completion dialog.
Actually, `:enhanced` and `:scrollbar` are not used.
I was also unsure whether to keep them or not, but even if they were there, IRB wouldn't malfunction so I left it. But should it be deleted?

Reply:
It looks like IRB's document dialog (which does not specify face yet) use it.

Reply:
Thanks for letting me know. 
Does it mean that a `default` value is used if the `face` argument is not provided in `Reline::DialogRenderInfo.new`?

Reply:
https://github.com/ruby/reline/pull/552/files#diff-082d93e1b70331eead4465512d8e0f47b9a91734b7c899d5929168c620cf8025R65-R67

As a result of ensuring that the absence of `:enhanced` and `:scrollbar` attributes are not broken even if the user forgets the scrollbar attribute in the completion_dialog config, enhanced and scrollbar attributes are automatically created in the `:default` config, too.
This is kind of a compromise with the existing implementation, where the concept of "completion_dialog" basically belongs to the IRB, but the SGR settings are set on the Reline side.

Reply:
Sorry, I overlooked tompng's comment.
default config is used here https://github.com/ruby/reline/pull/552/files#diff-9703d9141e08b96c2d616930b207873b52dc9eb68b81223818fe250cdb0e927fR834

Totally I made confusing this thread, lol

Reply:
Thank you! I understood.